genrl/deep/common/trainer.py

```diff
@@ -1,516 +1,472 @@
+        Number of environments
+        """
+        return self.env.n_envs
+
+
+class OffPolicyTrainer(Trainer):
+    """
+    Off-Policy Trainer class
+
+    :param agent: Algorithm object
+    :param env: Environment
+    :param logger: Logger object
+    :param x_axis: Key plotted on x_axis
+    :param buffer: Buffer Object. Cannot be None for Off-policy
+    :param off_policy: Is the algorithm off-policy?
+    :param save_interval: Model to save in each of these many timesteps
+    :param render: Should the Environment render
+    :param max_ep_len: Max Episode Length
+    :param distributed: (Should distributed training be enabled?
+(To be implemented))
+    :param ckpt_log_name: Model checkpoint name
+    :param steps_per_epochs: Steps to take per epoch?
+    :param epochs: Total Epochs to train for
+    :param device: Device to train model on
+    :param log_interval: Log important params every these many steps
+    :param batch_size: Size of batch
+    :param seed: Set seed for reproducibility
+    :param deterministic_actions: Take deterministic actions during training.
+    :param warmup_steps: (Observe the environment for these many steps
+with randomly sampled actions to store in buffer.)
+    :param start_update: (Starting updating the policy after these
+many steps)
+    :param update_interval: Update model policies after number of steps.
+    :type agent: object
+    :type env: object
+    :type logger: object
+    :type x_axis: str
+    :type buffer: object
+    :type off_policy: bool
+    :type save_interval:int
+    :type render: bool
+    :type max_ep_len: int
+    :type distributed: int
+    :type ckpt_log_name: string
+    :type steps_per_epochs: int
+    :type epochs: int
+    :type device: string
+    :type log_interval: int
+    :type batch_size: int
+    :type seed: int
+    :type deterministic_actions: bool
+    :type warmup_steps: int
+    :type start_update: int
+    :type update_interval: int
+    """
+
+    def __init__(
+        self,
+        agent: Any,
+        env: Union[gym.Env, VecEnv],
+        log_mode: List[str] = ["stdout"],
+        x_axis: str = "timestep",
+        buffer: Union[Type[ReplayBuffer], Type[PrioritizedBuffer]] = None,
+        off_policy: bool = True,
+        save_interval: int = 0,
+        render: bool = False,
+        max_ep_len: int = 1000,
+        distributed: bool = False,
+        ckpt_log_name: str = "experiment",
+        steps_per_epoch: int = 4000,
+        epochs: int = 10,
+        device: Union[torch.device, str] = "cpu",
+        log_interval: int = 10,
+        evaluate_episodes: int = 500,
+        logdir: str = "logs",
+        batch_size: int = 50,
+        seed: Optional[int] = 0,
+        deterministic_actions: bool = False,
+        warmup_steps: int = 1000,
+        start_update: int = 1000,
+        update_interval: int = 50,
+    ):
+        super(OffPolicyTrainer, self).__init__(
+            agent,
+            env,
+            log_mode,
+            x_axis,
```

Comment:
After merging, you'll have to use `x_axis`=`x_axis` here. Otherwise, the order in which the arguments are sent to the super class can be changed.